app/jobs/application_job.rb

```diff
@@ -6,8 +6,28 @@ class ApplicationJob < ActiveJob::Base
     exception.define_singleton_method(:skip_bugsnag) { true }
     raise exception
   end
+
   rescue_from ActiveRecord::Deadlocked, &skip_bugnag_and_raise
-  rescue_from ActiveJob::DeserializationError, &skip_bugnag_and_raise
+  rescue_from ActiveJob::DeserializationError do |exception|
+    # We except this to be a record not found. If it's not
+    # then get out of here
+    skip_bugnag_and_raise.(exception) unless exception.cause.is_a?(ActiveRecord::RecordNotFound)
+
+    # Let any transactions finish then look up again
+    # If the thing is now found, requeue the job.
+    # Or if it's still missing, just kill the job. This happens if
+    # the record is deleted in the time it takes to execute the job
+    # (e.g. a User's records being deleted)
+    sleep(0.5)
+    begin
+      exception.cause.model.constantize.find(exception.cause.id)
+      retry_job
+    rescue NoMethodError
+      # The model has gone. Kill the job.
+    rescue ActiveRecord::RecordNotFound
+      # The record is gone. Kill the job.
+    end
+  end
```

Comment:
This bit is a bit complex. We're looking at three issues:
1. The record is deleted (the final AR::RecordNotFound). In which case we just silently end the job.
2. The model is deleted (ie we've removed the file, which is the NoMethodError). In which case we just silently end the job.
3. The job has run before the db transaction has finished, in which case we sleep for half a second and then retry it.